tests/study_tests/test_multi_objective.py

```diff
@@ -0,0 +1,121 @@
+from typing import Sequence
+
+import pytest
+
+import optuna
+from optuna.study import StudyDirection
+from optuna.study._multi_objective import _dominates
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+def _create_trial(values: Sequence[float], state: TrialState = TrialState.COMPLETE) -> FrozenTrial:
+    return optuna.trial.create_trial(values=list(values), state=state)
+
+
+@pytest.mark.parametrize(
+    ("v1", "v2"), [(-1, 1), (-float("inf"), 0), (0, float("inf")), (-float("inf"), float("inf"))]
+)
+def test_dominates_1d_not_equal(v1: float, v2: float) -> None:
+    t1 = _create_trial([v1])
+    t2 = _create_trial([v2])
+
+    assert _dominates(t1, t2, [StudyDirection.MINIMIZE])
+    assert not _dominates(t2, t1, [StudyDirection.MINIMIZE])
+
+    assert _dominates(t2, t1, [StudyDirection.MAXIMIZE])
+    assert not _dominates(t1, t2, [StudyDirection.MAXIMIZE])
+
+
+@pytest.mark.parametrize("v", [0, -float("inf"), float("inf")])
+@pytest.mark.parametrize("direction", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+def test_dominates_1d_equal(v: float, direction: StudyDirection) -> None:
+    assert not _dominates(_create_trial([v]), _create_trial([v]), [direction])
+
+
+def test_dominates_2d() -> None:
+    directions = [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE]
+
+    # Check all pairs of trials consisting of these values, i.e.,
+    # [-inf, -inf], [-inf, -1], [-inf, 1], [-inf, inf], [-1, -inf], ...
+    # These values should be specified in ascending order.
+    vals = [-float("inf"), -1, 1, float("inf")]
+
+    # The following table illustrates an example of dominance relations.
+    # "D" cells in the table dominates the "t" cell in (MINIMIZE, MAXIMIZE) setting.
+    #
+    #                       v[1]
+    #      ╔═════╤═════╤═════╤═════╤═════╗
+    #      ║     │ -∞  │ -1  │  1  │  ∞  ║
+    #      ╟─────┼─────┼─────┼─────┼─────╢
+    #      ║ -∞  │     │     │  D  │  D  ║
+    #      ╟─────┼─────┼─────┼─────┼─────╢
+    #      ║ -1  │     │     │  D  │  D  ║
+    # v[0] ╟─────┼─────┼─────┼─────┼─────╢
+    #      ║  1  │     │     │  t  │  D  ║
+    #      ╟─────┼─────┼─────┼─────┼─────╢
+    #      ║  ∞  │     │     │     │     ║
+    #      ╚═════╧═════╧═════╧═════╧═════╝
+    #
+    # In the following code, we check that for each position of "t" cell, the relation
+    # above holds.
+
+    # Generate the set of all possible indices.
+    all_indices = set((i, j) for i in range(len(vals)) for j in range(len(vals)))
+    for (t_i, t_j) in all_indices:
+        # Generate the set of all indices that dominates the current index.
+        dominating_indices = set(
+            (D_i, D_j) for D_i in range(t_i + 1) for D_j in range(t_j, len(vals))
```

Comment:
A nit, but could you keep the coding convention?
We may also update the table above.
https://peps.python.org/pep-0008/#function-and-variable-names
```suggestion
            (d_i, d_j) for d_i in range(t_i + 1) for d_j in range(t_j, len(vals))
```